src/commissioner/Discoverer.h

```diff
@@ -0,0 +1,134 @@
+/*
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <commissioner/Commissionee.h>
+#include <controller/AbstractDnssdDiscoveryController.h>
+#include <setup_payload/SetupPayload.h>
+
+namespace chip {
+namespace Commissioner {
+namespace CommissionableNodeDiscoverer {
+
+struct Delegate
+{
+    virtual ~Delegate()                 = default;
+    virtual void OnDiscovery()          = 0;
+    virtual void OnDiscovererShutdown() = 0;
+};
+
+struct ShutdownToken
+{
+    ShutdownToken(Delegate ** delegate) : mDelegate(delegate) {}
+    ~ShutdownToken() { (*mDelegate)->OnDiscovererShutdown(); }
+
+private:
+    Delegate ** mDelegate;
+};
+
+struct Joinable
+{
+    Joinable(Delegate ** delegate) : mDelegate(delegate) {}
+
+protected:
+    bool InProgress() { return mShutdownToken.get() != nullptr ? true : false; }
+    void RetainShutdownToken(Platform::SharedPtr<ShutdownToken> token) { mShutdownToken = token; }
+    void ReleaseShutdownToken() { mShutdownToken = std::shared_ptr<ShutdownToken>(nullptr); }
```

Comment:
`mShutdownToken` is typed as `Platform::SharedPtr`, using `std::shared_ptr` may not fall in line with the spirit of Platform level types.

Reply:
I thought that since we are extensively using unique_ptr in our commissioning code, this was OK.

I guess it is OK to open up this conversation.  But I think that the utility and correctness imparted by the use of shared_ptr is worth a lot.

Reply:
I have fixed the use of std::shared_ptr (replaced with Platform::SharedPtr), and think the explanation that UniquePtr is already in use is reasonable justification that Platform::SharedPtr can also exist and be used in the project.